array.hpp

```diff
@@ -0,0 +1,282 @@
+        return array::const_iterator(_arr);//as I know , I will never attempt to change it's value , if it is really const
+    }
+
+    template<typename T, std::size_t N>
+    typename array<T,N>::iterator array<T, N>::end() {
+        return array::iterator(_arr+N);
+    }
+
+    template<typename T, std::size_t N>
+    typename array<T,N>::const_iterator array<T, N>::end()const {
+        return cend();
+    }
+
+    template<typename T, std::size_t N>
+    typename array<T,N>::const_iterator array<T, N>::cend()const {
+        return array::const_iterator(_arr);//as I know , I will never attempt to change it's value , if it is really const
+    }
+
+    template<typename T, std::size_t N>
+    typename array<T,N>::reverse_iterator array<T, N>::rbegin() {
+        return reverse_iterator(end());
+    }
+
+    template<typename T, std::size_t N>
+    typename array<T,N>::const_reverse_iterator array<T, N>::rbegin()const{
+        return crbegin();
+    }
+
+    template<typename T, std::size_t N>
+    typename array<T,N>::const_reverse_iterator array<T, N>::crbegin()const{
+        return const_reverse_iterator(cend());
+    }
+
+    template<typename T, std::size_t N>
+    typename array<T,N>::reverse_iterator array<T, N>::rend() {
+        return reverse_iterator(begin());
+    }
+
+    template<typename T, std::size_t N>
+    typename array<T,N>::const_reverse_iterator array<T, N>::rend()const{
+        return crend();
+    }
+
+    template<typename T, std::size_t N>
+    typename array<T,N>::const_reverse_iterator array<T, N>::crend()const{
+        return const_reverse_iterator(cbegin());
+    }
+
+    template<typename T, std::size_t N>
+    constexpr array<T, N>::array(const array &other) {
+        for(size_type i = 0; i < N; ++i){
+            _arr[i] = other[i];
+        }
+    }
+
+    template<typename T, std::size_t N>
+    constexpr array<T, N>::array(array &&other) {
+        for(size_type i = 0; i < N; ++i){
+            _arr[i] = std::move(other[i]);
+        }
+    }
+
+    template<typename T, std::size_t N>
+    constexpr array<T,N>& array<T, N>::operator = (const array & other) {
+        if(this != &other) {
+            for (size_type i = 0; i < N; ++i) {
+                _arr[i] = other[i];
+            }
+        }
+        return *this;
+    }
+
+    template<typename T, std::size_t N>
+    constexpr array<T,N>& array<T, N>::operator = (array && other) {
+        if(this != &other) {
+            for (size_type i = 0; i < N; ++i) {
+                _arr[i] = std::move(other[i]);
+            }
+        }
+        return *this;
+    }
+
+    template<typename T, std::size_t N>
+    constexpr array<T, N>::array(const std::initializer_list<T> &lst) {
+        if(lst.size() < N){
+            throw std::domain_error("the size of list can't be smaller than that of array's");
+        }
+        size_type index = 0;
+       for (const auto& i:lst){
+           if(index < N){
```

Comment:
if index < N ?

Reply:
Yes , why are you asking??

Reply:
`if(lst.size() < N){`
in the begining this check is done, isn't it enough?